# Lab 03-1: Operators and Strings

In [120]:
import math
import datetime

## Operators

### Exercise 1

***We’ve seen that n = 42 is legal. What about 42 = n?***

In [8]:
42 = n

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (3090476294.py, line 1)

`42 = n` is not legal as it is attempting to store `n` into the number `42`, which does not make sense and results in a syntax error. In other words, it attempts to assign a value to a constant (`42`). Constants (like literal numbers) cannot be assigned values; only variables can hold values. 

***How about x = y = 1?***

In [12]:
x = y = 1
print("x =", x)
print("y =", y)

x = 1
y = 1


`x` is assigned the value of `y`, which is assigned the value of `1`. Therefore, both `x` and `y` take on the value of `1`. The expression `x = y = 1` is an example of chained assignment. It means that both `x` and `y` are assigned the value `1` in a single line. This is a convenient way to assign the same value to multiple variables at once. In other words, we are assigning the same value to multiple variables in one line.

***In some languages every statement ends with a semi-colon, ;. What happens if you put a semi-colon at the end of a Python statement?***

In [16]:
x = y = 1;

In Python, adding a semicolon (;) at the end of a statement doesn't cause any issues—it is allowed but not required.

In [19]:
x = y = 1.

In [21]:
x = y = 1 .

SyntaxError: invalid syntax (724143125.py, line 1)

If we write `x = y = 1.`, the period indicates that `1.` is a floating-point number (`1.0`). This is valid and will assign the value `1.0` to `x`.

If we add a period that does not serve a purpose (e.g., `x = y = 1 .`), Python will raise a syntax error, because the period is not expected in this context.

We cannot place a period at the end of a Python statement unless it has a valid function, like representing a decimal number or being part of an object attribute.

***In math notation you can multiply x and y like this: x y. What happens if you try that in Python?***

In [27]:
x y

SyntaxError: invalid syntax (1567273689.py, line 1)

`x y` results in a syntax error as the expression is incomplete. Python does not understand what it means as there is no operator connecting the two variables. 

### Exercise 2

***1. The volume of a sphere with radius r is $\frac{4}{3} \pi r^3$. What is the volume of a sphere with radius 5?***

In [50]:
(4/3)*(math.pi)*(5)**3

523.5987755982989

The volume of a sphere with radius 5 is **524**.

***2. Suppose the cover price of a book is \\$24.95, but bookstores get a 40% discount. Shipping costs $3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?***

In [85]:
0.6*24.95*60+3+0.75*59

945.4499999999999

The total wholesale cost for 60 copies is **\\$945.45**.

***3. If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast?***

In [151]:
start_time = datetime.datetime(1,1,1,6,52,0) # dummy date 1, 1, 1
print("Start time:", start_time.time())
seconds_taken = (8*60+15)+3*(7*60+12)+(8*60+15)
print("Seconds taken =", seconds_taken)
end_time = start_time + datetime.timedelta(seconds = seconds_taken)
print("End time:", end_time.time())

Start time: 06:52:00
Seconds taken = 2286
End time: 07:30:06


You will get home for breakfast at **7:30 am**.

## Strings

### Exercise 1

***Read the documentation of the string methods at http://docs.python.org/3/library/stdtypes.html#string-methods. You might want to experiment with some of them to make sure you understand how they work. `strip` and `replace` are particularly useful.***

In [212]:
# str.replace("a", "b"): replacing all occurrences
text = "I love apples, apples are great."
new_text = text.replace("apples", "oranges")
print(new_text)

# str.replace("a", "b", n): limiting number of replacements
text = "I love apples, apples are great."
new_text = text.replace("apples", "oranges", 1) # count parameter is set to 1
print(new_text)

# str.strip(char): removes any combination of the characters specified in the string passed to the strip() method from both the start and the end of the string. 
# It does not remove characters from the middle of the string.
text = "I love apples."
new_text = text.strip('I love .')
print(new_text)

# str.rstrip(char): returns a copy of the string with trailing characters removed. 
text = "I love apples."
new_text = text.rstrip('apples.')
print(new_text)

# str.lstrip(char): returns a copy of the string with leading characters removed. 
text = "I love apples."
new_text = text.lstrip('I love ')
print(new_text)

I love oranges, oranges are great.
I love oranges, apples are great.
apples
I love 
apples.


### Exercise 2

***There is a string method called `count` that is similar to the function in Section 8.7. Read the documentation of this method and write an invocation that counts the number of `a`’s in `'banana'`.***

In [251]:
# str.count(sub[, start[, end]]): returns the number of non-overlapping occurrences of substring sub in the range [start, end].
text = "banana"
count_a = text.count('a')
print(count_a)

3


The number of `a`'s in `'banana'` is **3**.

### Exercise 3

***A string slice can take a third index that specifies the “step size”; that is, the number of spaces between successive characters. Use this idiom to write a one-line version of `is_palindrome` from Exercise 3.***

In [279]:
def is_palindrome(word): return word == word[::-1] # [::-1] creates a reversed version of the string

print(is_palindrome('allen'))
print(is_palindrome('bob'))
print(is_palindrome('otto'))
print(is_palindrome('redivider'))

False
True
True
True


### Exercise 4

***The following functions are all intended to check whether a string contains any lowercase letters, but at least some of them are wrong. For each function, describe what the function actually does (assuming that the parameter is a string).***

In [348]:
def any_lowercase1(s):
    for c in s:
        if c.islower():
            return True
        else:
            return False

print(any_lowercase1("I have an apple"))

False


This function only checks whether the **first** character of the string is a lowercase letter.

In [351]:
def any_lowercase2(s):
    for c in s:
        if 'c'.islower():
            return 'True'
        else:
            return 'False'

print(any_lowercase2("I have an apple"))

True


`'c'` is being interpreted as the literal string `'c'` (a single character string) rather than the variable c. Therefore, `'c'.islower()` is always checking whether the literal character `'c'` is lowercase, which is always `'True'`.

In [360]:
def any_lowercase3(s):
    for c in s:
        flag = c.islower()
    return flag

print(any_lowercase3("I have an apple"))
print(any_lowercase3("I have an applE"))

True
False


The function overwrites the value of `flag` on each iteration of the loop. By the time the loop finishes, `flag` will only hold the result (`True` or `False`) for the last character in the string, thereby ignoring the rest of the characters that come before it.

In [371]:
def any_lowercase4(s):
    flag = False
    for c in s:
        flag = flag or c.islower()
    return flag

print(any_lowercase4("I have an apple"))
print(any_lowercase4("I HAVE AN APPLE"))
print(any_lowercase4("I have an applE"))

True
False
True


The function is technically correct, though slightly inefficient. Once the function detects a lowercase letter, it continues to check the remaining characters, even though the result is already determined (i.e., `flag` will remain `True`).

In [386]:
def any_lowercase5(s):
    for c in s:
        if not c.islower():
            return False
    return True

print(any_lowercase5("ihaveanapple"))
print(any_lowercase5("i have an apple."))
print(any_lowercase5("I HAVE AN APPLE"))
print(any_lowercase5("I have an applE"))

True
False
False
False


The function will return `False` as soon as it encounters any character that is **not** lowercase. In other words, the function checks whether ***all*** characters are lowercase, not whether there are ***any*** lowercase characters. 

Any space or punctuation mark (i.e., `.`) will also result in a `False` output even if all the other characters are lowercase.

### Exercise 5

***Write a function called `rotate_word` that takes a string and an integer as parameters, and returns a new string that contains the letters from the original string rotated by the given amount.***

In [421]:
def rotate_word(word, n): # rotate a word by n places (word = string, n = integer)
    result = '' # initialise an empty string `result` to store the rotated characters

    for c in word:
        # check if the character is a letter
        if c.isalpha():
            # determine if the character is uppercase or lowercase
            start = ord('A') if c.isupper() else ord('a')
            # rotate character within the alphabet
            rotated_c = chr(start + (ord(c) - start + n) % 26)
            result += rotated_c
        else:
            # if it is not a letter, leave it unchanged
            result += c
    
    return result

print(rotate_word("abc", 1)) 
print(rotate_word("XYZ", 5))
print(rotate_word("IBM", -1))

bcd
CDE
HAL
